Remove SchemaAdapter

adriangb opened the pull request:
Closes #16800

We could leave some of these methods around as deprecated and make them no-ops but I'd be afraid that would create a false sense of security (compiles but behaves wrong at runtime).

alamb commented on docs/source/library-user-guide/upgrading.md:
```diff
@@ -490,6 +490,24 @@ If you were using a custom `SchemaAdapterFactory` for schema adaptation (e.g., d
+- `SchemaMapping` struct
+- `DefaultSchemaAdapterFactory` struct
+
+These types were previously used to adapt record batch schemas during file reading.
```
This is likely to cause non trivial pain for anyone who uses the SchemaAdapter during upgrade

However, I am not sure if leaving the code in but disconnected would be any better.

Thus I think we should go with this PR and we can help with some more writeups when we start testing the upgrade with downstream crates (like delta.rs)

adriangb replied:
How about I leave them in, mark them as deprecated and have them raise a runtime error with a link to the upgrading guide? At least then it doesn't fail silently at runtime. You'd have to ignore the compile time warnings as well.

alamb reviewed (approved):
Thank you @adriangb 

I think this PR is good enough to go. I do think it would be worth porting some of the tests as well to .slt files and @martin-g 's suggestion to consolidate the upgrade guide

But I think it is important to close out the expression rewriting changes in DataFusion 52 and get this done

kosiew commented on datafusion/datasource/src/file.rs:
```diff
@@ -157,30 +156,4 @@ pub trait FileSource: Send + Sync {
-    fn with_schema_adapter_factory(
-        &self,
-        _factory: Arc<dyn SchemaAdapterFactory>,
-    ) -> Result<Arc<dyn FileSource>> {
```
The FileSource trait no longer has with_schema_adapter_factory() and schema_adapter_factory() methods.

For users, this means there's now no way for a custom FileSource to influence schema adaptation behavior at the file-source level. The only knob is at FileScanConfig / ListingTableConfig, which is downstream.

This is a capability reduction that should be called out in migration docs.

adriangb replied:
Can you give an example of a use case for these trait methods? Who is using a custom `FileSource` without using `FileScanConfig`? Why can't they attach the adapter to the concrete type, i.e. why does it have to be part of the trait? As far as I could see these methods were only being used by our own tests.

kosiew reviewed (commented):
<img width="173" height="55" alt="Image" src="https://github.com/user-attachments/assets/9c859be6-8e79-477e-a324-adabfce6ca4b" />

Thanks for the amazing work!

adriangb commented on datafusion/core/tests/schema_adapter/schema_adapter_integration_tests.rs:
```diff
@@ -1,752 +0,0 @@
-#[tokio::test]
-#[cfg(feature = "parquet")]
-async fn test_parquet_flipped_projection() -> Result<()> {
-    // Create test data with columns (a, b, c) - the file schema
```
I believe we have several but I added an slt version in 2f5ff2e796a291da63ec1c8484fe73d295ffba6b

adriangb commented:
I think I've addressed all of the feedback except maybe https://github.com/apache/datafusion/pull/19345#discussion_r2626740331. Since I've already had to resolve conflicts a couple times and it seems this would hold up a release I'd like to merge this approved PR. @kosiew is that concern a blocker or can we discuss more and address as a followup?
